libp2p/security/insecure/transport.py

```diff
@@ -21,12 +46,29 @@ class InsecureTransport(BaseSecureTransport):
         for an inbound connection (i.e. we are not the initiator)
         :return: secure connection object (that implements secure_conn_interface)
         """
-        return InsecureSession(self, conn, ID(b""))
+        session = InsecureSession(self, conn, ID(b""))
+        await session.run_handshake()
+        return session
 
     async def secure_outbound(self, conn: IRawConnection, peer_id: ID) -> ISecureConn:
         """
         Secure the connection, either locally or by communicating with opposing node via conn,
         for an inbound connection (i.e. we are the initiator)
         :return: secure connection object (that implements secure_conn_interface)
         """
-        return InsecureSession(self, conn, peer_id)
+        session = InsecureSession(self, conn, peer_id)
+        await session.run_handshake()
+        # TODO: Check if `remote_public_key is not None`. If so, check if `session.remote_peer`
+        received_peer_id = session.get_remote_peer()
+        if session.get_remote_peer() != peer_id:
+            raise UpgradeFailure(
+                "remote peer sent unexpected peer ID. "
+                f"expected={peer_id} received={received_peer_id}"
+            )
```

Comment:
Should we close the session in this case?

Reply:
IMO we should leave the `Connection.close` to the caller. Will do it in `Swarm.conn_handler`.